src/hammer-vlsi/par/openroad/__init__.py

```diff
@@ -80,152 +197,1018 @@ def export_config_outputs(self) -> Dict[str, Any]:
+
+        # pin_access - this command caused openroad to crash
+        # -allow_congestion and -allow_overflow added for now
+
+        global_route -guide_file {self.route_guide_path} -allow_congestion -congestion_iterations 200 -verbose
+
+        set_propagated_clock [all_clocks]
+        estimate_parasitics -global_routing
+
+        check_antennas -report_file {self.run_dir}/antenna.log -report_violating_nets
+        """)
         return True
 
-    def opt_design(self) -> bool:
-        # TODO: currently no analagous OpenROAD default script
+    def detailed_route(self) -> bool:
+        metals=self.get_stackup().metals[1:]
+        
+        self.block_append(f"""
+        ################################################################
+        # Detailed routing
+
+        # set_thread_count [exec getconf _NPROCESSORS_ONLN]
+
+        write_verilog -include_pwr_gnd {self.output_netlist_filename}
+
+        # TODO: many other arguments available
+        detailed_route -guide {self.route_guide_path} \\
+            -bottom_routing_layer met1 \\
+            -top_routing_layer met5 \\
+            -output_guide {self.run_dir}/{self.top_module}_output_guide.mod \\
+            -output_drc {self.run_dir}/{self.top_module}_route_drc.rpt \\
+            -output_maze {self.run_dir}/{self.top_module}_maze.log \\
+            -verbose 1 \\
+            -droute_end_iter 3
+
+        #utl::metric "DRT::drv" [detailed_route_num_drvs]
+        #write_def {self.run_dir}/{self.top_module}_route.def
+        """)
+        return True
+
+    def extraction(self) -> bool:
+        sed_expr=r"{s/\\//g}"  # use sed find+replace to remove '\' character
+        self.block_append(f"""
+        ################################################################
+        # Extraction
+        define_process_corner -ext_model_index 0 X
+
+        extract_parasitics -ext_model_file {self.get_setting("par.inputs.openrcx_techfile")}
+
+        # touch the file in case write_spef fails
+        exec touch {self.output_spef_paths}
+        write_spef {self.output_spef_paths}
+        # remove backslashes in instances so that read_spef recognizes the instances
+        exec sed -i {sed_expr} {self.output_spef_paths}
+
+        read_spef {self.output_spef_paths}
+        """)
+        # alternative: use global routing based parasitics inlieu of rc extraction
+        #   >> estimate_parasitics -global_routing
+        return True
+
+    def write_netlist(self) -> bool:
+        self.append(f"write_verilog -include_pwr_gnd -remove_cells {{ {self.fill_cells} }} {self.output_netlist_filename}")
+        return True
+
+    # Copy and hack the klayout techfile, to add all required LEFs
+    def setup_klayout_techfile(self) -> bool:
+        setting_dir = self.get_setting("technology.sky130.sky130A")
+        setting_dir = Path(setting_dir)
+
+        source_path = setting_dir / 'libs.tech' / 'klayout' / 'sky130A.lyt'
```

Comment:
This source_path seems sky130 specific - but isn't it specified in the tech json?

Reply:
You're right, fixed.